x-pack/plugins/beats/server/routes/api/register_assign_tags_to_beats_route.js

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Joi from 'joi';
+import {
+  get,
+  flatten,
+  uniq
+} from 'lodash';
+import { INDEX_NAMES } from '../../../common/constants';
+import { callWithRequestFactory } from '../../lib/client';
+import { wrapEsError } from '../../lib/error_wrappers';
+
+async function getDocs(callWithRequest, ids) {
+  const params = {
+    index: INDEX_NAMES.BEATS,
+    type: '_doc',
+    body: { ids },
+    _source: false
+  };
+
+  const response = await callWithRequest('mget', params);
+  return get(response, 'docs', []);
+}
+
+function getBeats(callWithRequest, beatIds) {
+  const ids = beatIds.map(beatId => `beat:${beatId}`);
+  return getDocs(callWithRequest, ids);
+}
+
+function getTags(callWithRequest, tags) {
+  const ids = tags.map(tag => `tag:${tag}`);
+  return getDocs(callWithRequest, ids);
+}
+
+async function findNonExistentItems(callWithRequest, items, getFn) {
+  const itemsFromEs = await getFn.call(null, callWithRequest, items);
+  return itemsFromEs.reduce((nonExistentItems, itemFromEs, idx) => {
+    if (!itemFromEs.found) {
+      nonExistentItems.push(items[idx]);
+    }
+    return nonExistentItems;
+  }, []);
+}
+
+function findNonExistentBeatIds(callWithRequest, beatIds) {
+  return findNonExistentItems(callWithRequest, beatIds, getBeats);
+}
+
+function findNonExistentTags(callWithRequest, tags) {
+  return findNonExistentItems(callWithRequest, tags, getTags);
+}
+
+async function persistAssignments(callWithRequest, assignments) {
+  const body = flatten(assignments.map(assignment => {
+    const { beatId, tag } = assignment;
+    const script = 'def beat = ctx._source.beat; '
+      + 'if (beat.tags == null) { '
+      + '  beat.tags = []; '
+      + '} '
+      + 'if (!beat.tags.contains(params.tag)) { '
+      + '  beat.tags.add(params.tag); '
+      + '}';
```

Comment:
@nik9000 Do you have any suggestions for this script? I'm using it in a bulk API call, doc update operation (see a couple lines below here where I construct the bulk API request body "rows").

My docs in ES have a field called `beat`, which has a sub-field called `tags`. Here I'm trying to insert a new tag w/o duplication in the list of tags.

Reply:
This *should* be fine, just understand that if you have many tags this will start to get slow because these are lists and `contains` is linear.

I'd probably put the `def beat` string on the some indentation as the first `+ '` line so a casual reader like myself doesn't miss it. But I dunno what the language allows here.

Reply:
Noted re: O(n) `contains`. We don't expect the number of tags for a beat to grow beyond dozens (@exekias?) but I don't suppose there's a way around the linear time of `contains`, given that there's no notion of a Set datatype in ES, right?

Reply:
I don't think that will be a common case, in general I think number of tags should be ~1-6, of course with exceptions.

Reply:
Perfect, thanks @exekias. In that case I'm not going to worry about the performance over here for now.